services/meta/data.go

```diff
@@ -783,6 +784,91 @@ func (data *Data) hasAdminUser() bool {
 	return false
 }
 
+// ImportData imports selected data into the current metadata.
+// opts provides options to possibly rename the Database or Retention Policy, or change the Replication Factor.
+func (data *Data) ImportData(other Data, backupDBName, restoreDBName, backupRPName, restoreRPName string) (map[uint64]uint64, error) {
+
+	if backupDBName == "" {
+		// if no backupDBName then we'll try to import all the DB's.  If one of them fails, we'll mark the whole
+		// operation a failure and return an error.
+		shardIDMap := make(map[uint64]uint64)
+		for _, dbi := range other.Databases {
+			if dbi.Name == "_internal" {
+				continue
+			}
+			tmpMap, err := data.ImportOneDB(other, dbi.Name, "", "", "")
+			if err != nil {
+				return shardIDMap, err
+			}
+
+			for k, v := range tmpMap {
+				shardIDMap[k] = v
+			}
+		}
+		return shardIDMap, nil
+	}
+
+	return data.ImportOneDB(other, backupDBName, restoreDBName, backupRPName, restoreRPName)
+}
+
+// ImportOneDB imports a single database/rp from an external metadata object, renaming them if new names are provided.
+func (data *Data) ImportOneDB(other Data, backupDBName, restoreDBName, backupRPName, restoreRPName string) (map[uint64]uint64, error) {
+
+	shardIDMap := make(map[uint64]uint64)
+
+	dbImport := other.Database(backupDBName)
+	if dbImport == nil {
+		return shardIDMap, fmt.Errorf("imported metadata does not have datbase named %s", backupDBName)
+	}
+
+	if restoreDBName == "" {
+		restoreDBName = backupDBName
+	}
+
+	// change the names if we want/need to
+	dbImport.Name = restoreDBName
+
+	if backupRPName != "" {
+		rpImport, err := other.RetentionPolicy(restoreDBName, backupRPName)
+		if err != nil {
+			return shardIDMap, err
+		}
+
+		if rpImport != nil {
+			if restoreRPName == "" {
+				restoreRPName = backupRPName
+			}
+			rpImport.Name = restoreRPName
+			dbImport.RetentionPolicies = []RetentionPolicyInfo{*rpImport}
+		} else {
+			return shardIDMap, fmt.Errorf("retention Policy not found in meta backup: %s", backupRPName)
+		}
+
+	}
+	if data.Database(restoreDBName) != nil {
+		return shardIDMap, errors.New("database already exists")
+	}
+
+	for _, rpImport := range dbImport.RetentionPolicies {
```

Comment:
I would strongly recommend updating this to create new info objects throughout this code, rather than mutating what's already here. The metadata structs are full of slices, so you may be inadvertently modifying shared data.

Reply:
solved using dbInfo.clone()  which does a deep copy.